### PR TITLE
[MAINTENANCE] Add support for Typo3 v12 (fixes #900)

### LINKED ISSUE #900: [FUND] Support for TYPO3 12 LTS
## Description

The upcoming Kitodo.Presentation 5.0 release will support TYPO3 10 ELTS and 11 LTS. The next [TYPO3 LTS version](https://typo3.org/cms/roadmap) is already released:

* [TYPO3 12 LTS](https://get.typo3.org/)

The next major release 6.x should support TYPO3 11 LTS and 12 LTS.

## Expected benefits of this development
* Developer / Maintainer
   * This is mandatory. We cannot ignore the development of the upstream TYPO3 CMS. It pushes the development and makes sure Kitodo.Presentation is an up-to-date tool.
* Administrators
   * Run supported TYPO3-version is obligatory.

## Estimated Costs and Complexity

I assume this should be a **medium cost** project for TYPO3 12, since there are just a few breaking changes to be expected. But the minimum PHP version will increase from 7.4 to 8.1, which requires some changes. Since TYPO3 11 LTS is supported until 2024-10-31, this is IMHO a must-do for the development fund 2024.

### PR DESCRIPTION
**Pull request by thomaslow**
### Changes

The following changes have been implemented:

- Update composer.json for Typo3 v12
- Add missing dependency for `typo3/cms-scheduler`
- Add version restriction for phpunit v9 (phpunit v10 has breaking changes)
- Upgrade dependency `typo3/testing-framework` to v7, which is compatible with Typo3 v11 and v12
- Fix basic problems with extension configuration (see Typo3 [Feature #96733](https://docs.typo3.org/c/typo3/cms-core/main/en-us/Changelog/12.0/Feature-96733-NewBackendModuleRegistrationAPI.html))
- Replace `TYPO3_MODE` references (see Typo3 [Deprecation #92947](https://docs.typo3.org/c/typo3/cms-core/main/en-us/Changelog/11.0/Deprecation-92947-DeprecateTYPO3_MODEAndTYPO3_REQUESTTYPEConstants.html))
- Remove phpstan ignore rule `TYPO3_MODE not found`, which is not required any more
- Fix flash message rendering (see Typo3 [Depcrecation #97787](https://docs.typo3.org/c/typo3/cms-core/main/en-us/Changelog/12.0/Deprecation-97787-SeveritiesOfFlashMessagesAndReportsDeprecated.html))
- Fix `SolrSearch` implementation by extending from `Query` class (see Typo3 [Breaking #96044](https://docs.typo3.org/c/typo3/cms-core/main/en-us/Changelog/12.0/Breaking-96044-HardenMethodSignatureOfLogicalAndAndLogicalOr.html))
- Avoid initialization via constructor in `AbstractController` (see Typo3 [Deprecation #99615](https://docs.typo3.org/c/typo3/cms-core/main/en-us/Changelog/12.2/Deprecation-99615-GeneralUtilityGPMerged.html]))
- Fix NewTenantController by using dependency injection (see Typo3 [Manual](https://docs.typo3.org/m/typo3/reference-coreapi/11.5/en-us/ApiOverview/DependencyInjection/Index.html#when-to-use-dependency-injection-in-typo3))
- Return ResponseInterfaces in controller actions (see Typo3 [Deprecation #92784](https://docs.typo3.org/c/typo3/cms-core/main/en-us/Changelog/11.0/Deprecation-92784-ExtbaseControllerActionsMustReturnResponseInterface.html))
- Change `forward` calls to `redirect` in controller classes (see Typo3 [Deprecation #92815](https://docs.typo3.org/c/typo3/cms-core/main/en-us/Changelog/11.0/Deprecation-92815-ActionControllerForward.html)) 
- Remove reference to old `ObjectManager` (see Typo3 [Deprecation #94619](https://docs.typo3.org/c/typo3/cms-core/main/en-us/Changelog/11.4/Deprecation-94619-ExtbaseObjectManager.html))
- Determine login status via Context class instead of `$GLOBALS["TSFE"]` (see Typo3 [Manual](https://docs.typo3.org/m/typo3/reference-coreapi/11.5/en-us/ApiOverview/Context/Index.html))
- Update GitHub CI tests to Typo3 v11 (with PHP 7.4) and v12 (with PHP 8.1)
- Remove columns `cruser_id`, `t3ver_oid`, `t3ver_wsid`, `t3ver_state`, `t3ver_stage` from test fixtures, which are not available in Typo3 v12 any more (see Typo3 [Breaking #98024](https://docs.typo3.org/c/typo3/cms-core/main/en-us/Changelog/12.0/Breaking-98024-TCA-option-cruserid-removed.html))
- Fix OAI-PMH template in case of request without verb or other parameters
- Fix various undefined array key issues that are raised when using PHP 8+
- Fix `session_id` length in `tx_dlf_basket` table, which has more then 32 characters since Typo3 v10.4, see  [ext_tables.sql](https://github.com/TYPO3/typo3/blob/808ef61a53bc1e7ef65b715c13727b0f2d086c24/typo3/sysext/frontend/ext_tables.sql#L29) for Typo3 frontend module.
- Fix url parameter not added in navigation and toc links in Typo3 v12, see Typo3 [Feature #98488](https://docs.typo3.org/c/typo3/cms-core/main/en-us/Changelog/12.0/Feature-98488-AdditionalSettingForTypolinkOptionAddQueryString.html)

### Todo

The following tasks still need to be done before merging:

- [x] Remove restriction for `solr:9.6` as soon as pull request #1335 is merged
- [x] Revert fixes for PHP 8.1 as soon as pull request #1332 is merged
- [x] Further testing, see below

### Tests

The following tests have been done for a fresh installation of both Typo3 v11 and Typo3 v12:

General
- [x] Unit and functional tests
- [x] Triggering functions `Create default namespaces`, `Create structures`, `Create metadata`, `Create solr rore` on `New Tenant` admin page
- [x] Index several documents from URLs of METS files via `kitodo:index`
- [x] Re-index all documents via `kitodo:reindex --all`
- [x] Harvest documents from a OAI baseurl via `kitodo:harvest`
- [x] Create new Typo3 pages and add Kitodo Plugins
- [x] Test output of Kitodo Plugins with `kitodo-demo` theme, see below

Kitodo Plugins
- [x] `Kitodo: Collection`
  - has problems: same in master
- [x] `Kitodo: Search` 
  - seems to work: forwards to page containing `Kitodo: List View`, shows facets, facets can be selected
- [x] `Kitodo: List View` 
  - seems to work: shows search results, forwards to page containing `Kitodo: PageView`
- [x] `Kitodo: PageView` 
  - seems to work: image visible, can be zoomed
- [x] `Kitodo: Navigation` 
  - seems to work: shows navigation elements, other pages can be selected
- [x] `Kitodo: Feeds` 
  - seems to work: shows RSS feed with documents
- [x] `Kitodo: OAI-PMH Interface` 
  - has problems: shows XSLT transformed OAI-PMH responses, but no records (same in master)
- [x] `Kitodo: Table of Contents` 
  - seems to work: toc is shown, sections can be selected
- [x] `Kitodo: Basket` 
  - seems to work: add document to basket, can show basket
- [x] `Kitodo: Metadata` 
  - seems to work: shows metadata of document in page view
- [x] `Kitodo: Page Grid` 
  - seems to work: shows page thumbnails, can navigate to page
- [x] `Kitodo: Calendar` 
  - seems to work: shows years, shows months, newspaper issue can be selected

#### Unable to test
- `Kitodo: Statistics` 
- `Kitodo: Embedded 3d Viewer` 
- `Kitodo: AudioPlayer` 
- `Kitodo: Toolbox` 
- `Kitodo: Basket` (pdf generation)

Unfortunately, I do not have access or experience with a production system of `kitodo-presentation`. Because of that, I'm not sure whether some of the issues are the result of a misconfiguration of my Typo3 installation. All problems listed above also occur with the current master and my Typo3 installation. I would appreciate any help in testing my changes or tips to correctly set up Typo3 pages and Kitodo-Presentation plugins for testing.

### Demo

Demonstration of `kitodo-presentation` updated for Typo3 v12 with demo theme (please ignore CSS problems)

https://github.com/user-attachments/assets/03de016a-9ef4-444c-a52e-28ab6d7add3c

### CODE REVIEW

**Comment by sebastian-meyer** [comment]
Hm, some parent classes method's signature for some of the classes in `Classes/Validation` have changed from TYPO3 v11 to v12. So the tests for v12 fail, because our classes don't explicitly declare the same parameter and return types as the TYPO3's parent classes. But if I add the type declarations, the tests for v11 fail, because now their signatures don't match with the parent classes in v11.
Any idea how we could avoid that? Maybe don't use `declare(strict_types=1);` in those classes?

**Comment by sebastian-meyer** [comment]
Also pinging @markusweigelt, because he is the author of the validator feature.

**Comment by thomaslow** [comment]
@sebastian-meyer Thanks for trying to merge. I'll have to look at this in more detail. It may be related to [Breaking Change 96998](https://docs.typo3.org/c/typo3/cms-core/main/en-us/Changelog/12.0/Breaking-96998-ExtbaseValidatorInterfaceChanged.html#breaking-96998-extbase-validator-interface-changed).

**Comment by markusweigelt** [comment]
Thank you for your implementations. I will test the 'Embedded 3D Viewer' later this week with your update. 

Regarding the validation: 
Should we support both versions? It seems that security releases for TYPO3 11.5 are ending this month.

https://docs.typo3.org/c/typo3/cms-core/main/en-us/Changelog/12.0/Breaking-96998-ExtbaseValidatorInterfaceChanged.html#migration has some hint how to handle both versions.

**Comment by sebastian-meyer** [comment]
Yes, if it is possible we should support both versions. There are some users who will use the ELTS version of TYPO3 v11 before migrating to v12.

I reverted my changes in `master`. All validation classes are now again working for v10 and v11, but not for v12.

**Comment by thomaslow** [comment]
@sebastian-meyer Unit and integration tests work again with both Typo3 v11 and v12. Please let me know in case I can help some more.

**Comment by sebastian-meyer** [comment]
Thank you very much! I'll review this tomorrow!

**Review by sebastian-meyer — APPROVED** [review]
Outstanding work! Thank you very much!

I'll properly separate branches for 5.x and 6.x development and then this can be merged.

**Comment by thomaslow** [comment]
@sebastian-meyer I'm glad my changes work for you. Please let me know in case any issues pop up that I missed. Of course, I'll occasionally check the issue list myself in the following weeks.

File: Classes/Domain/Repository/TokenRepository.php [hunk 1]
```diff
@@ -42,7 +42,7 @@ public function deleteExpiredTokens(int $expireTime): void
         $constraints[] = $query->lessThan('tstamp', (int) (time() - $expireTime));
 
         if (count($constraints)) {
-            $query->matching($query->logicalAnd($constraints));
+            $query->matching($query->logicalAnd(...array_values($constraints)));
```
Comment (beatrycze-volk) [inline Classes/Domain/Repository/TokenRepository.php:45]: @thomaslow What is the reason behind this change? `$constraints` is already an array so this change looks not necessary.

Comment (thomaslow) [reply Classes/Domain/Repository/TokenRepository.php:45]: @beatrycze-volk The `...` (unpacking of array) change was necessary due to the API change described in [Breaking #96044](https://docs.typo3.org/c/typo3/cms-core/main/en-us/Changelog/12.0/Breaking-96044-HardenMethodSignatureOfLogicalAndAndLogicalOr.html). I am not sure whether `array_values()` is really necessary.

Comment (beatrycze-volk) [reply Classes/Domain/Repository/TokenRepository.php:45]: Thanks for answer. I would suspect that only `...` are needed, as array is already declared.

Comment (beatrycze-volk) [reply Classes/Domain/Repository/TokenRepository.php:45]: As per documentation `array_values` function returns `array`:
array_values([array](https://www.php.net/manual/en/language.types.array.php) $array): [array](https://www.php.net/manual/en/language.types.array.php)

Then I will remove call of `array_values` and make PR for it.
